File: atcoder/abc293/d.py
N, M = map(int, input().split())

adj = [[] for _ in range(2 * N)]
for i in range(N):
    adj[2 * i + 0].append(2 * i + 1)
    adj[2 * i + 1].append(2 * i + 0)

for _ in range(M):
    a, b, c, d = input().split()
    u = (int(a) - 1) * 2 + (1 if b == 'B' else 0)
    v = (int(c) - 1) * 2 + (1 if d == 'B' else 0)
    adj[u].append(v)
    adj[v].append(u)

cnt1, cnt2 = 0, 0
vis = [False for _ in range(2 * N)]
for root in range(2 * N):
    if vis[root]:
        continue

    stack = [root]
    verts = set([root])
    while len(stack) > 0:
        u = stack.pop()
        for v in adj[u]:
            if not vis[v]:
                vis[v] = True
                stack.append(v)
                verts.add(v)

    # print(root, verts, [len(adj[v]) for v in verts])
    if all(len(adj[v]) % 2 == 0 for v in verts):
        cnt1 += 1
    else:
        cnt2 += 1

print(cnt1, cnt2)

File: atcoder/abc293/d.rs
#![allow(unused)]

use std::collections::HashSet;

fn main() {
    let inp = readv::<usize>();
    let (n, m) = (inp[0], inp[1]);
    let mut adj = vec![vec![]; 2 * n];
    for i in 0..n {
        adj[2 * i + 0].push(2 * i + 1);
        adj[2 * i + 1].push(2 * i + 0);
    }
    for _ in 0..m {
        let inp = readv::<String>();
        let (a, b, c, d) = (
            inp[0].clone(),
            inp[1].clone(),
            inp[2].clone(),
            inp[3].clone(),
        );
        let a: usize = a.parse().unwrap();
        let c: usize = c.parse().unwrap();
        let u = (a - 1) * 2 + if b == "R" { 0 } else { 1 };
        let v = (c - 1) * 2 + if d == "R" { 0 } else { 1 };
        adj[u].push(v);
        adj[v].push(u);
    }

    let mut vis = vec![false; 2 * n];
    let mut cnt1 = 0;
    let mut cnt2 = 0;
    for root in 0..2 * n {
        if vis[root] {
            continue;
        }

        let mut stack = vec![root];
        let mut verts = HashSet::new();
        verts.insert(root);
        while let Some(u) = stack.pop() {
            for &v in adj[u].iter() {
                if !vis[v] {
                    vis[v] = true;
                    stack.push(v);
                    verts.insert(v);
                }
            }
        }

        if verts.iter().all(|&v| adj[v].len() % 2 == 0) {
            cnt1 += 1;
        } else {
            cnt2 += 1;
        }
    }

    println!("{} {}", cnt1, cnt2);
}

fn read<T: std::str::FromStr>() -> T {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).ok();
    s.trim().parse().ok().unwrap()
}

fn readv<T: std::str::FromStr>() -> Vec<T> {
    read::<String>()
        .split_ascii_whitespace()
        .map(|t| t.parse().ok().unwrap())
        .collect()
}

fn reads() -> Vec<char> {
    read::<String>().chars().collect::<Vec<char>>()
}

fn join<T: ToString>(arr: &[T], sep: &str) -> String {
    arr.iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join(sep)
}
